sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,416 @@
+                raise ValueError(msg)
+
+        param_grid = dict(
+            covariance_type=covariance_type,
+            n_components=range(self.min_components, self.max_components + 1),
+        )
+        param_grid = list(ParameterGrid(param_grid))
+
+        seeds = random_state.randint(np.iinfo(np.int32).max, size=len(param_grid))
+
+        if parse_version(joblib.__version__) < parse_version("0.12"):
+            parallel_kwargs = {"backend": "threading"}
+        else:
+            parallel_kwargs = {"prefer": "threads"}
+
+        results = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, **parallel_kwargs)(
+            delayed(self._fit_cluster)(X, gm_params, seed)
+            for gm_params, seed in zip(param_grid, seeds)
+        )
+        best_criter = [result.criterion for result in results]
+
+        if sum(best_criter == np.min(best_criter)) == 1:
+            best_idx = np.argmin(best_criter)
+        else:
+            # in case there is a tie,
+            # select the model with the least number of parameters
+            ties = np.where(best_criter == np.min(best_criter))[0]
+            n_params = [results[tie].model._n_parameters() for tie in ties]
+            best_idx = ties[np.argmin(n_params)]
+
+        self.best_criterion_ = results[best_idx].criterion
+        self.n_components_ = results[best_idx].n_components
+        self.covariance_type_ = results[best_idx].covariance_type
+        self.best_model_ = results[best_idx].model
+        self.n_iter_ = results[best_idx].model.n_iter_
+        self.labels_ = results[best_idx].model.predict(X)
+        self.results_ = results
+        self.n_features_in_ = X.shape[1]
+
+        return self
+
+    def predict(self, X):
+        """Predict clusters based on the best Gaussian mixture model.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points. Each row
+            corresponds to a single data point.
+
+        Returns
+        -------
+        labels : array, shape (n_samples,)
+            Component labels.
+        """
+        check_is_fitted(self, ["best_model_"], all_or_any=all)
+        X = self._validate_data(X, reset=False)
+        labels = self.best_model_.predict(X)
+
+        return labels
+
+    def fit_predict(self, X, y=None):
+        """Fit the models and predict clusters based on the best model.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points. Each row
+            corresponds to a single data point.
+
+        y : Ignored
+            Not used, present for API consistency by convention.
+
+        Returns
+        -------
+        labels : array, shape (n_samples,)
+            Component labels.
+        """
+        self.fit(X, y)
+
+        labels = self.predict(X)
+        return labels
+    
+
+
+class _CollectResults:
```

Comment:
this is effectively a dictionary - recommend just using one, or a named tuple? i am just anti classes that only store data and dont have any methods, but that is just my style :)